source/keyboardHandler.py

```diff
@@ -219,7 +219,19 @@ def internal_keyDownEvent(vkCode,scanCode,extended,injected):
 			for k in range(256):
 				keyStates[k]=ctypes.windll.user32.GetKeyState(k)
 			charBuf=ctypes.create_unicode_buffer(5)
+			# First try getting the keyboard layout from the thread with the focus (input thread)
 			hkl=ctypes.windll.user32.GetKeyboardLayout(focus.windowThreadID)
+			if not hkl:
+				log.debug("Failed to fetch keyboard layout from focus, trying layout from last detected change")
+				# Some threads, such as for Windows consoles
+				# Do not allow getKeyboardLayout to work.
+				# Therefore, use the cached keyboard layout from the last inputLangChange detected by NVDA
+				# on the foreground object.
+				hkl = getattr(api.getForegroundObject(), '_lastDetectedKeyboardLayoutChange', 0)
+				if not hkl:
+					log.debug("No layout cached, falling back to layout of NVDA main thread")
+					# As a last resort, use the keyboard layout of NVDA's main thread.

```

Comment:
How likely would it be that the keyboard layout for the NVDA main thread differs from the keyboard layout of the currently focused app?